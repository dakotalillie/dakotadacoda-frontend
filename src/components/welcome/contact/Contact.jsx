import React from 'react';
import {
  Button,
  Col,
  Container,
  Form,
  FormGroup,
  Input,
  Label,
  Row
} from 'reactstrap';

import { determineButtonClass, determineButtonIcon } from './utils';
import './contact.css';

export default class Contact extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      formValues: {
        name: '',
        email: '',
        _subject: '',
        message: '',
      },
      formErrors: {
        name: false,
        email: false,
        _subject: false,
        message: false,
      },
      status: null,
      errorMessage: '',
      buttonHover: false
    };
    this.form = null;

    this.setFormRef = e => {
      this.form = e;
    };
  }

  findEmptyFields = () => {
    const emptyFields = [];
    for (const field of Object.keys(this.state.formValues)) {
      if (this.state.formValues[field].length === 0)
        emptyFields.push(field);
    }
    if (emptyFields.length > 0) {
      const newFormErrors = { ...this.state.formErrors };
      emptyFields.forEach(field => {
        newFormErrors[field] = true;
      });
      this.setState({
        formErrors: newFormErrors,
        status: 'error',
        errorMessage: 'Every field must be filled in.'
      });
    }
    return emptyFields;
  }

  handleChange = e => {
    const name = e.target.name;
    const value = e.target.value;
    this.setState(prevState => {
      const newState = { ...prevState };
      newState.formValues[name] = value;
      return newState;
    });
  };

  handleFocus = e => {
    const name = e.target.name;
    if (this.state.formErrors[name]) {
      const newFormErrors = { ...this.state.formErrors };
      newFormErrors[name] = false;
      this.setState({ formErrors: newFormErrors, status: null });
    }
  }

  handleSubmit = e => {
    e.preventDefault();
    if (!this.state.status || this.state.status === 'error') {
      const emptyFields = this.findEmptyFields();
      if (!emptyFields.length) {
        this.setState({ status: 'loading' }, () => {
          this.sendEmail();
        });
      }
    }
  };

  sendEmail = () => {
    fetch('https://formspree.io/mkreevwx', {
      method: 'POST',
      body: JSON.stringify(this.state.formValues),
      headers: {
        'Content-Type': 'application/json'
      }
    }).then(this.handleResponse);
  }

  handleResponse = res => {
    if (res.status === 200) {
      this.setState({ status: 'success' });
    } else {
      this.setState({
        status: 'error',
        errorMessage: 'Oops! Something went wrong. Please try again later.'
      });
    }
  }

  render() {
    return (
      <div className="contact section">
        <Container>
          <Row>
            <Col>
              <h2>Get In Touch</h2>
            </Col>
          </Row>
          <Form
            className="contact_form"
            onSubmit={this.handleSubmit}
            innerRef={this.setFormRef}
          >
            <Row>
              <Col sm="6">
                <FormGroup>
                  <Label for="name">Name</Label>
                  <Input
                    type="text"
                    name="name"
                    id="name"
                    placeholder="Jane Doe"
                    className={this.state.formErrors['name'] ? 'error' : ''}
                    value={this.state.formValues.name}
                    onChange={this.handleChange}
                    onFocus={this.handleFocus}
                  />
                </FormGroup>
                <FormGroup>
                  <Label for="email">Email</Label>
                  <Input
                    type="email"
                    name="email"
                    id="email"
                    placeholder="your@email.com"
                    className={this.state.formErrors['email'] ? 'error' : ''}
                    value={this.state.formValues.email}
                    onChange={this.handleChange}
                    onFocus={this.handleFocus}
                  />
                </FormGroup>
                <FormGroup>
                  <Label for="_subject">Subject</Label>
                  <Input
                    type="text"
                    name="_subject"
                    id="subject"
                    placeholder="Your Subject"
                    className={this.state.formErrors['_subject'] ? 'error' : ''}
                    value={this.state.formValues._subject}
                    onChange={this.handleChange}
                    onFocus={this.handleFocus}
                  />
                </FormGroup>
              </Col>
              <Col sm="6">
                <FormGroup>
                  <Label for="message">Message</Label>
                  <Input
                    type="textarea"
                    name="message"
                    id="message"
                    placeholder="Hello World!"
                    className={this.state.formErrors['message'] ? 'error' : ''}
                    value={this.state.formValues.message}
                    onChange={this.handleChange}
                    onFocus={this.handleFocus}
                  />
                </FormGroup>
              </Col>
            </Row>
            {this.state.status === 'error'
              ? <p className="error_text">{this.state.errorMessage}</p>
              : null
            }
            <div className="button_div">
              <Button
                outline
                size="lg"
                color="primary"
                className={determineButtonClass(this.state.status)}
                type="submit"
                onMouseEnter={() => this.setState({ buttonHover: true })}
                onMouseLeave={() =>
                  this.setState({
                    buttonHover: false
                  })
                }
              >
                {determineButtonIcon(this.state.status, this.state.buttonHover)}
              </Button>
            </div>
          </Form>
        </Container>
      </div>
    );
  }
}
