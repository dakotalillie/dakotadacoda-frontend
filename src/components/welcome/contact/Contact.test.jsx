import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Contact from './Contact';

describe('Contact', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Contact />);
    wrapper.instance().sendEmail = jest.fn();
    wrapper.update();
  });

  it('renders successfully', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('user fills in some fields', () => {
    beforeEach(() => {
      wrapper.find('#name').simulate('change', {
        target: { name: 'name', value: 'Dakota' }
      });
      wrapper.find('#email').simulate('change', {
        target: { name: 'email', value: 'dakota@email.com' }
      });
    });

    it('updates state', () => {
      expect(wrapper.state().formValues.name).toEqual('Dakota');
      expect(wrapper.state().formValues.email).toEqual('dakota@email.com');
    });

    describe('user submits form with empty fields', () => {

      beforeEach(() => {
        wrapper.find('Form').simulate('submit', { preventDefault: () => { } });
      });

      it('updates state', () => {
        expect(wrapper.state().formErrors.subject).toBe(true);
        expect(wrapper.state().formErrors.text).toBe(true);
        expect(wrapper.state().status).toEqual('error');
      });

      it('determines input className from state', () => {
        expect(wrapper.find('#subject').props().className).toEqual('error');
        expect(wrapper.find('#text').props().className).toEqual('error');
      });

      it('clears error after focusing on input', () => {
        wrapper.find('#subject').simulate('focus', { target: { name: 'subject' }});
        expect(wrapper.state().formErrors.subject).toBe(false);
        expect(wrapper.state().status).toBeNull();
      });

    });

    describe('user fills in rest of fields and submits form', () => {

      beforeEach(() => {
        wrapper.find('#subject').simulate('change', {
          target: { name: 'subject', value: 'Foo Bar' }
        });
        wrapper.find('#text').simulate('change', {
          target: { name: 'text', value: 'Hello World' }
        });
        wrapper.find('Form').simulate('submit', { preventDefault: () => { } });
      });    
  
      it('sets status to `loading`', () => {
        expect(wrapper.state().status).toEqual('loading');
      });
  
      it('calls sendEmail', () => {
        expect(wrapper.instance().sendEmail).toHaveBeenCalled();
      });

      it('changes submit button className', () => {
        const button = wrapper.find('.button_div Button');
        expect(button.props().className).toEqual('submit_button disabled loading');
      });

      it('changes submit button contents', () => {
        const button = wrapper.find('.button_div Button');
        expect(button.children().text()).toEqual('<Loader />');
      });

      describe('submission is successful', () => {

        beforeEach(() => {
          wrapper.instance().handleResponse({ status: 200 });
          wrapper.update();
        });

        it('sets status to `success`', () => {   
          expect(wrapper.state().status).toEqual('success');
        });

        it('changes submit button className', () => {
          const button = wrapper.find('.button_div Button');
          expect(button.props().className).toEqual('submit_button disabled');
        });
  
        it('changes submit button contents', () => {
          const button = wrapper.find('.button_div Button');
          expect(button.children().text()).toEqual('<FontAwesomeIcon /> Sent!');
        });
      });

      describe('submission is unsuccessful', () => {

        beforeEach(() => {
          wrapper.instance().handleResponse({ status: 400 });
          wrapper.update();
        });

        it('sets status to `error`', () => {
          expect(wrapper.state().status).toEqual('error');
        });
      
        it('displays error message ', () => {
          expect(wrapper.find('.error_text').exists()).toBe(true);
        });

        it('resets submit button className', () => {
          const button = wrapper.find('.button_div Button');
          expect(button.props().className).toEqual('submit_button');
        });
  
        it('resets submit button contents', () => {
          const button = wrapper.find('.button_div Button');
          expect(button.children().text()).toEqual('Send');
        });
      });
    });
  });
});