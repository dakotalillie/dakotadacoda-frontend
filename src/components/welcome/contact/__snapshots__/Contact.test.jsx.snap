// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact renders successfully 1`] = `
<div
  className="contact section"
>
  <Container
    tag="div"
  >
    <Row
      tag="div"
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <h2>
          Get In Touch
        </h2>
      </Col>
    </Row>
    <Form
      className="contact_form"
      innerRef={[Function]}
      onSubmit={[Function]}
      tag="form"
    >
      <Row
        tag="div"
      >
        <Col
          sm="6"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <FormGroup
            tag="div"
          >
            <Label
              for="name"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              Name
            </Label>
            <Input
              className=""
              id="name"
              name="name"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Jane Doe"
              type="text"
              value=""
            />
          </FormGroup>
          <FormGroup
            tag="div"
          >
            <Label
              for="email"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              Email
            </Label>
            <Input
              className=""
              id="email"
              name="email"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="your@email.com"
              type="email"
              value=""
            />
          </FormGroup>
          <FormGroup
            tag="div"
          >
            <Label
              for="subject"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              Subject
            </Label>
            <Input
              className=""
              id="subject"
              name="subject"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Your Subject"
              type="text"
              value=""
            />
          </FormGroup>
        </Col>
        <Col
          sm="6"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <FormGroup
            tag="div"
          >
            <Label
              for="text"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              Message
            </Label>
            <Input
              className=""
              id="text"
              name="text"
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Hello World!"
              type="textarea"
              value=""
            />
          </FormGroup>
        </Col>
      </Row>
      <div
        className="button_div"
      >
        <Button
          className="submit_button"
          color="primary"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          outline={true}
          size="lg"
          tag="button"
          type="submit"
        >
          Send
        </Button>
      </div>
    </Form>
  </Container>
</div>
`;
